### ISSUE TITLE
Error stream in loss functions

### ISSUE THREAD

**Issue by favre49** [mlpack/mlpack#2346]
Currently, metrics in `src/mlpack/core/cv/metrics` handle errors like this:

```
if (data.n_cols != responses.n_cols)
{
        std::ostringstream oss;
         oss << "MSE::Evaluate(): number of points (" << data.n_cols << ") "
            << "does not match number of responses (" << responses.n_cols << ")!"
            << std::endl;
       throw std::invalid_argument(oss.str());
 }
```
However, the rest of the codebase used `Log::Fatal`. To establish uniformity across the codebase, we should probably change that.

**Comment by kartikdutt18** [comment]
Hey @bisakhmondal, Taking this up right?

**Comment by bisakhmondal** [comment]
I have made the necessary changes.

**Comment by kartikdutt18** [comment]
Great. Looking forward to your PR.

**Comment by bisakhmondal** [comment]
Just curious to know, is there any performance difference between `Log::fatal ` and throwing exception using std because I found there is also a lot of files is handling the same using the 2nd way.
for eg.(a few)
`../mlpack/src/mlpack/methods/softmax_regression/softmax_regression.cpp
../mlpack/src/mlpack/methods/lsh/lsh_search_impl.hpp
../mlpack/src/mlpack/methods/linear_svm/linear_svm_impl.hpp
../mlpack/src/mlpack/methods/lmnn/lmnn_function_impl.hpp
../mlpack/src/mlpack/methods/range_search/range_search_impl.hpp
../mlpack/src/mlpack/methods/ann/augmented/tasks/copy_impl.hpp
../mlpack/src/mlpack/methods/ann/augmented/tasks/score_impl.hpp
../mlpack/src/mlpack/methods/ann/augmented/tasks/sort_impl.hpp
../mlpack/src/mlpack/methods/ann/augmented/tasks/add_impl.hpp
../mlpack/src/mlpack/methods/decision_tree/decision_tree_impl.hpp
`

**Comment by rcurtin** [comment]
No, I don't think there's any significant performance difference (though to be honest when we're at the point of throwing an exception, performance isn't a concern anymore since something has clearly gone wrong :smile:).  `Log::Fatal` didn't always previously throw an exception (in earlier mlpack versions, it actually called `exit(1)`...); so my guess here is that when that code was written, it was done to avoid the call to `exit(1)`, and then when `Log::Fatal` was changed to throw an exception instead of calling `exit(1)`, this code wasn't changed.  Personally I'd be fine changing it all to use `Log::Fatal` now.

See also #891 for some more related discussion.

**Comment by bisakhmondal** [comment]
> though to be honest when we're at the point of throwing an exception, performance isn't a concern anymore since something has clearly gone wrong 

Exactly!! It totally got slipped from my mind :). Thanks @rcurtin. Got it. I will make the necessary changes.

**Comment by favre49** [comment]
Yeah no issues with that, that makes sense

**Comment by bisakhmondal** [comment]
Yeah @zoq, totally understand your concern.

> we need to see how we can give the user interpretable exceptions that they can actually do something with. Remember that if an exception is thrown, then typically you'll only get some short error message like matrix of incorrect size or failed to converge or something like this; that's not enough for a user to actually understand what has gone wrong. That is the bigger issue that we need to solve here.

From the #891 discussion, I would like to know more about the idea of interpretable exception.
Thanks.

**Comment by rcurtin** [comment]
Ahh, sorry, I realize I keep waffling on this.  In #891 I said that the better idea would be to use specific exception types.  There's one other advantage to that approach: if we use `Log::Fatal`, things are *always* printed to `stderr`.  But this isn't always useful for users; it's primarily useful from bindings.

If I'm a user writing C++ code that uses mlpack, it might be really irritating to have things dump to `[FATAL]` on stderr even when they are exceptions that I am catching and handling.  (I'd then have to dig into the mlpack sources and find out how to disable it, which is possible but probably still irritating for users.)

Sorry that I didn't think this through originally.  It's been a long time since #891 has been discussed, but it would be nice to perhaps continue the discussion there?  Maybe we can come to some kind of conclusion.  It seems like there are really two issues there:

 1. *What do we use internally for error reporting?*  (this seems to be the primary focus of this issue)  I might say now, that after thinking about it, in the bindings, we can call `Log::Fatal`, and outside of the bindings, we should throw typed exception.  That would mean that any code calling `Log::Fatal` outside of a `_main.cpp` file (or a file that's only included by a `_main.cpp` file) should be adapted to throw properly typed exceptions instead.

 2. *How can we provide _useful_ error messages to users?*  (this seems out of scope for here, but I still think it would be cool to be able to provide some kind of nicer backtrace support when there's a failure.  That would also be a big deal for notebook debugging :)  But anyway, maybe we should pick that up in #891 if there's momentum to discuss it now?)

I looked at #2348, thank you @bisakhmondal for taking the time to implement it! :) Even if we don't merge it, one nice thing in there is that you discovered the function `AssertSizes()`, which I didn't know about.  I think maybe we could use that more in the mlpack codebase and it can make things cleaner (perhaps we should move it to `core/util/` and rename it `CheckSameSizes()`; we could also add one called `CheckSameDimensionality()`).  Maybe that's a separate issue, we could open another one if you like.

Hope my comments are helpful, sorry again for suggesting one thing and then thinking about it and going with another approach. :-O

**Comment by bisakhmondal** [comment]
It's completely fine @rcurtin, for the previous suggestion at least I got to dig more into the codebase.

> If I'm a user writing C++ code that uses mlpack, it might be really irritating to have things dump to [FATAL] on stderr even when they are exceptions that I am catching and handling. 

Yup, agree with you. It's pretty irritating every time to see a `FATAL` message even when an exception is being caught.
The 2 issues, you have nicely depicted, I can make the necessary changes in the current issue for the first one.
For the second one, I will continue the discussion on #891( maybe 1-2 days later), once I became familiarized with the present backtrace utility.
The `CheckSameSizes()`, `CheckSameDimensionality()` utilities will be nice to be integrated into the codebase as it will significantly reduce the code size. Feel free to open an issue, I'll take care of that.

Thank you @rcurtin for such an informative comment. So can I make the changes here for the error reporting part?

**Comment by rcurtin** [comment]
I'd say maybe it's better to wait a few days to see if anyone else has any input before making changes, but yeah, overall, I'd be happy if you made the changes for whatever we agree on. :+1:

**Comment by bisakhmondal** [comment]
Yeah, Sure!! :+1:

**Comment by kartikdutt18** [comment]
Let's Keep this open.

**Comment by kartikdutt18** [comment]
Keep Open.
